src/codegen/events/mod.rs

```diff
@@ -0,0 +1,47 @@
+mod solana;
+mod substrate;
+
+use crate::codegen::cfg::ControlFlowGraph;
+use crate::codegen::events::solana::SolanaEventEmitter;
+use crate::codegen::events::substrate::SubstrateEventEmitter;
+use crate::codegen::vartable::Vartable;
+use crate::codegen::Options;
+use crate::sema::ast;
+use crate::sema::ast::{Function, Namespace};
+use crate::Target;
+use solang_parser::pt;
+
+/// This traits delineates the common behavior of event emission. As each target uses a different
+/// encoding scheme, there must be an implementation of this trait for each.
+pub(super) trait EventEmitter {
+    /// Generate the CFG instructions for emitting an event
+    fn emit(
```

Comment:
Did you consider to return a `Result` instead? Unless there is absolutely no potential errors here for (which I doubt)?

Reply:
All the necessary code analysis is dune during the parsing and semantic analysis. If the compiler enters code generation, we assume it is possible to generate the CFG for the AST we've been given. There is no need to return an error here.

Reply:
Alright, I added a comment to point that out. With that I think it makes for a reasonable interface.